src/mme-app/actionHandlers/attachActionHandlers.cpp

```diff
@@ -1444,3 +1484,102 @@ ActStatus ActionHandlers::handle_state_guard_timeouts_for_csreq_ind(ControlBlock
     mme_tables->initiate_spgw_resolution();
     return ActStatus::ABORT;
 }
+/***************************************
+* Action handler : req_to_dns
+***************************************/
+ActStatus ActionHandlers::req_to_dns(ControlBlock& cb)
+{
+	log_msg(LOG_DEBUG, "Inside req_to_dns \n");
+	 if(mme_cfg->dns_config.dns_flag == 1)
+        {
+
+	uint8_t plmn_id[3] = {0};
+
+        UEContext *ue_ctxt = dynamic_cast<UEContext*>(cb.getPermDataBlock());
+        if (ue_ctxt == NULL)
+        {
+                log_msg(LOG_DEBUG, "send_air_to_hss: ue context is NULL \n");
+                return ActStatus::HALT;
+        
+	}
+
+	 MmeDnsMsgUtils::SetDNSConfiguration();
+	/* Query DNS based on lb and hb of tac */
+	char lb[LB_HB_LEN] = {0};
+	char hb[LB_HB_LEN] = {0};
+	char mnc[MCC_MNC_LEN] = {0};
+	char mcc[MCC_MNC_LEN] = {0};
+	short tac=0;
+	char mccDigit1 =  ue_ctxt->getTai().tai_m.plmn_id.idx[0] & 0x0F;
+	char mccDigit2 = (ue_ctxt->getTai().tai_m.plmn_id.idx[0] & 0xF0) >> 4;
+	char mccDigit3 = ue_ctxt->getTai().tai_m.plmn_id.idx[1] & 0x0F;
+	char mncDigit1 = ue_ctxt->getTai().tai_m.plmn_id.idx[2] & 0x0F;
+	char mncDigit2 = (ue_ctxt->getTai().tai_m.plmn_id.idx[2] & 0xF0) >> 4;
+	char mncDigit3 = (ue_ctxt->getTai().tai_m.plmn_id.idx[1] & 0xF0) >> 4;
+
+
+
+	if (mncDigit3 == 15)
+			snprintf(mnc, MCC_MNC_LEN, "%u%u", mncDigit1,mncDigit2);
+	else
+			snprintf(mnc, MCC_MNC_LEN, "%u%u%u", mncDigit1,mncDigit2,mncDigit3);
+
+
+	snprintf(mcc,MCC_MNC_LEN, "%u%u%u", mccDigit1, mccDigit2, mccDigit3);
+
+	log_msg(LOG_DEBUG,"%u %u  %u %u %u %u mcc and mnc digits\n", mccDigit1, mccDigit2, mccDigit2, mncDigit1, mncDigit2, mncDigit3);
+	log_msg(LOG_DEBUG, "%s %s: mnc and mcc\n", mnc,mcc);
+	memcpy(plmn_id, ue_ctxt->getTai().tai_m.plmn_id.idx, 3);
+	log_msg(LOG_DEBUG, "%s %s: plmn_id\n", plmn_id,ue_ctxt->getTai().tai_m.plmn_id.idx);
+	tac = ntohs(ue_ctxt->getTai().tai_m.tac);
+	log_msg(LOG_DEBUG, "%d %d tac value* \n",tac, ue_ctxt->getTai().tai_m.tac);
+
+	if ( tac != 1) {
+			log_msg(LOG_DEBUG, "Could not get SGW-U list using DNS query. TAC missing in CSR.\n");
+			return ActStatus::HALT;
+	}
+
+	sprintf(lb, "%u", ue_ctxt->getTai().tai_m.tac & 0xFF);
+	sprintf(hb, "%u", (ue_ctxt->getTai().tai_m.tac >> 8) & 0xFF);
+	log_msg(LOG_DEBUG, "%s %s lb and hb value is\n", lb,hb);
+	
+	EPCDNS::SGWNodeSelector s2(lb, hb,mnc,mcc);
+        s2.setNamedServerID(DNS::NS_DEFAULT);
+        s2.addDesiredProtocol( EPCDNS::SGWAppProtocolEnum::sgw_x_s11 );
+	uint32_t ue_idx = ue_ctxt->getContextID();
+	s2.process(&ue_idx,NodeSelector_test_callback);
+	ProcedureStats::num_of_processed_dns_response ++;
+	}
+	
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+* Action handler : process_dns_resp
+***************************************/
+ActStatus ActionHandlers::process_dns_resp(ControlBlock& cb)
+{
+   if(mme_cfg->dns_config.dns_flag == 1)
```

Comment:
Should we check dns_flag config in the process_dns_resp.

Reply:
Yes, this should run only in case if dns_flag is true.